### ISSUE TITLE
Show prefix text in TextField even when it is not in focus and empty

### ISSUE THREAD

**Issue by shujames** [flutter/flutter#19488]
Currently there is no way to force prefix text defined in the InputDecoration to stay shown if text field is not in focus and text is empty. 
https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/input_decorator.dart#L1669

Is there a way to extend InputDecorator to allow that behavior.

**Comment by willlarche** [comment]
We were talking about this recently. Let me get a final call on it.

**Comment by willlarche** [comment]
We've spoken with Design. 

__tl:dr Let's allow this functionality with caveats.__

### Details

This is partially tested and partially untested behavior so the informal guidance will be as follows:

* Text-based prefixes are best when no wider than an icon.
  * When they are wider than an icon, the animation and the `x` value of the floating placeholder is undefined. The animation could end up crossing the prefix or the placeholder could end up too far trailing if it pins to the trailing of the prefix for its floating `x`.
* Text-based suffixes are untested and can't have any guidance right now. It could end up being confusing to have "@github.com" as a suffix. 
* Currency symbols are probably the best use case at the moment. 
* Any text-based prefix or suffix should be 1 line and baseline aligned to the input text. 
* Use all this at your own risk until formal testing has been carried out. It's not at all known when that could happen.

**Comment by Hixie** [comment]
So if I'm understanding the above correctly, we should have a flag on text fields that says whether or not the prefix and suffix should be included even when not focused and with no text.

cc @HansMuller

**Comment by mehmetf** [comment]
@HansMuller could you please let us know if the decorator is capable of doing what Mulligan is asking? If not, could we add this functionality as it is a launch blocker.

**Comment by mehmetf** [comment]
@willlarche is working on this. He will have an update in the week of August 13.

**Comment by willlarche** [comment]
@shujames I just wanna make sure I understand the original request. 

The prefix text is inline with the floating label when the label is down. If it's always showing, it will overlap:

![screen shot 2018-08-14 at 2 51 31 pm](https://user-images.githubusercontent.com/1271525/44111856-8de86082-9fd1-11e8-8248-c52b95056458.png)

If you want prefix text that is laid out to the leading of the label, that already exists with the `prefixIcon` field on `InputDecoration`. It takes any `Widget`. If you pass it a `Text` widget in a `Center` widget in a `SizedBox` widget, you'll get a layout that's pleasing:

![screen shot 2018-08-14 at 2 53 17 pm](https://user-images.githubusercontent.com/1271525/44111957-cd05bd50-9fd1-11e8-9924-f5382aeaf496.png)

**Comment by willlarche** [comment]
Closing for now as it's not possible to keep the prefix text visible in all states with a floating placeholder which is the main usage of this widget.

**Comment by noelmansour** [comment]
Maybe this should be a separate issue, but the problem with using `prefixIcon` is that it has a minimum size of 48x48 as can be seen in `input_decorator.dart`:

```dart
final Widget prefixIcon = decoration.prefixIcon == null ? null :
Center(
  widthFactor: 1.0,
  heightFactor: 1.0,
  child: ConstrainedBox(
    constraints: const BoxConstraints(minWidth: 48.0, minHeight: 48.0),
    child: IconTheme.merge(
      data: IconThemeData(
        color: iconColor,
        size: iconSize,
      ),
      child: decoration.prefixIcon,
    ),
  ),
);
```

This results in something like this:

<img width="492" alt="screen shot 2019-02-01 at 5 59 58 pm" src="https://user-images.githubusercontent.com/327631/52154430-c752b600-264b-11e9-9b42-fc9e29efcb22.png">

Whereas `prefixText` result in something that aligns much better:

<img width="491" alt="screen shot 2019-02-01 at 6 00 50 pm" src="https://user-images.githubusercontent.com/327631/52154365-7a6edf80-264b-11e9-9d46-356286eb8776.png">

`prefixIcon` code for reference:

```dart
TextField(
  decoration: InputDecoration(
    prefixIcon: SizedBox(
      child: Center(
        widthFactor: 0.0,
        child: Text('\$', style: Theme.of(context).textTheme.display1),
      ),
    ),
  ),
  style: Theme.of(context).textTheme.display1,
)
```

`prefixText` code:

```dart
TextField(
  decoration: InputDecoration(
    prefixText: '\$',
  ),
  style: Theme.of(context).textTheme.display1,
);
```

If `prefixIcon` is the approach to take, there should be a way to override the 48x48 constraint.

**Comment by serkon** [comment]
prefixIcon size must be changeable! 
Our designers will hate us if we can't change expected in the design...

**Comment by NoamDev** [comment]
I for example, need prefix text for the '+' in dial code.
If look at whatsapp registration page, you'll see that the plus is shown even when the field isn't in focus..

**Comment by NoamDev** [comment]
@willlarche what about making an option for always prefix text, but that it'll depend on null label?
That should bypass the problem of prefix text and label at the same time. because, icon is not applicable always, like the '+' in the dial code.

**Comment by NoamDev** [comment]
well, maybe #30664 can be the solution.

**Comment by PiN73** [comment]
A workaround:

```dart
TextStyle _decorationStyleOf(BuildContext context) {
  final theme = Theme.of(context);
  return theme.textTheme.subhead
      .copyWith(color: theme.hintColor);
}

Stack(
  alignment: Alignment.centerLeft,
  children: <Widget>[
    Builder(
      builder: (context) {
        return Text(
          'prefix',
          style: _decorationStyleOf(context),
        );
      }
    ),
    TextField(
      decoration: InputDecoration(
        prefixText: 'prefix',
        prefixStyle: TextStyle(color: Colors.transparent),
      ),
    ),
  ],
)
```

**Comment by droidluv** [comment]
@PiN73 thanks for the workaround

**Comment by patelkrunal78** [comment]
Very Similar To [How to add clear button to TextField Widget](https://icetutor.com/question/how-to-add-clear-button-to-textfield-widget)

**Comment by AxesandGrinds** [comment]
> A workaround:
> 
> ```dart
> TextStyle _decorationStyleOf(BuildContext context) {
>   final theme = Theme.of(context);
>   return theme.textTheme.subhead
>       .copyWith(color: theme.hintColor);
> }
> 
> Stack(
>   alignment: Alignment.centerLeft,
>   children: <Widget>[
>     Builder(
>       builder: (context) {
>         return Text(
>           'prefix',
>           style: _decorationStyleOf(context),
>         );
>       }
>     ),
>     TextField(
>       decoration: InputDecoration(
>         prefixText: 'prefix',
>         prefixStyle: TextStyle(color: Colors.transparent),
>       ),
>     ),
>   ],
> )
> ```

this doesn't work, it only makes the prefix always transparent.

**Comment by alejo9719** [comment]
I found a workaround for this, which is to set the prefixIcon (which is always shown) to a Text widget wrapped by a padding (because of alignment issues).

`prefixIcon: Padding(padding: EdgeInsets.all(15), child: Text('$prefixTextStr')),`
